### PR TITLE
refactor: switch over createWithTimestamps in LPG model

### PR DESCRIPTION
**Pull request by andreivladbrg**
Also:
- feat: add createWithTimestampsLPG in the SablierBatchLockup contract 
- test: remove duplicated tests from SafeOracle

Addresses:
- https://cantina.xyz/code/fed466ff-5256-4c6a-8021-6da2c113beb2/findings/1
- https://cantina.xyz/code/fed466ff-5256-4c6a-8021-6da2c113beb2/findings/10

### CODE REVIEW

File: lockup/src/interfaces/ISablierBatchLockup.sol [hunk 1]
```diff
@@ -99,6 +99,25 @@ interface ISablierBatchLockup {
...
+    /// @dev Requirements:
+    /// - There must be at least one element in `batch`.
+    /// - All requirements from {ISablierLockupPriceGated.createWithTimestampsLPG} must be met for each stream.
+    ///
+    /// @param lockup The address of the {SablierLockup} contract.
+    /// @param token The contract address of the ERC-20 token to be distributed.
+    /// @param batch An array of structs, each encapsulating a subset of the parameters of
+    /// {ISablierLockupPriceGated.createWithTimestampsLPG}.
+    /// @return streamIds The ids of the newly created streams.
+    function createWithTimestampsLPG(
```
Comment (smol-ninja) [inline lockup/src/interfaces/ISablierBatchLockup.sol:117]: Should we add a note explaining why only `createWithTimestampsLPG` is supported for LPG model? It can mention reasons such as contract size limit etc and we may support `createWithDurationsLPG` in the future when contract size is no longer an issue.

File: lockup/src/interfaces/ISablierLockupPriceGated.sol [hunk 2]
```diff
@@ -58,12 +57,10 @@ interface ISablierLockupPriceGated is ISablierLockupState {
     ///
     /// @param params Struct encapsulating the function parameters, which are documented in {Lockup} type.
     /// @param unlockParams Struct encapsulating the unlock parameters, documented in {LockupPriceGated}.
-    /// @param duration The total duration of the stream in seconds.
     /// @return streamId The ID of the newly created stream.
-    function createWithDurationsLPG(
-        Lockup.CreateWithDurations calldata params,
-        LockupPriceGated.UnlockParams calldata unlockParams,
-        uint40 duration
+    function createWithTimestampsLPG(
```
Comment (smol-ninja) [inline lockup/src/interfaces/ISablierLockupPriceGated.sol:63]: A similar note be added here too.

File: lockup/tests/integration/concrete/lockup-price-gated/create-with-timestamps-lpg/createWithTimestampsLPG.tree [hunk 3]
```diff
@@ -0,0 +1,10 @@
+CreateWithTimestampsLPG_Integration_Concrete_Test
```
Comment (smol-ninja) [inline lockup/tests/integration/concrete/lockup-price-gated/create-with-timestamps-lpg/createWithTimestampsLPG.tree:1]: Since there is only one function for LPG, what if we create a full BTT tree for this similar to what we had in `createWithDUrationsLPG.tree`?

The reason we had a short tree for other models is because they are well tested in the create durations function test

File: lockup/tests/integration/concrete/lockup-price-gated/streamed-amount-of/streamedAmountOf.t.sol [hunk 4]
```diff
@@ -6,7 +6,7 @@ import { Lockup_PriceGated_Integration_Concrete_Test } from "../LockupPriceGated
 contract StreamedAmountOf_Lockup_PriceGated_Integration_Concrete_Test is Lockup_PriceGated_Integration_Concrete_Test {
     function test_GivenDepletedStream() external {
         // Forward time to the end time.
-        vm.warp({ newTimestamp: getBlockTimestamp() + defaults.TOTAL_DURATION() });
+        vm.warp({ newTimestamp: defaults.END_TIME() + 1 });
```
Comment (smol-ninja) [inline lockup/tests/integration/concrete/lockup-price-gated/streamed-amount-of/streamedAmountOf.t.sol]: If it was not failing for only `defaults.END_TIME()`, should we just check for that? its an edge case and if it works for `defaults.END_TIME()`, it will work for `defaults.END_TIME() + 1`

File: lockup/tests/invariant/handlers/LockupCreateHandler.sol [hunk 5]
```diff
@@ -126,14 +126,14 @@ contract LockupCreateHandler is BaseHandler, Calculations {
         store.pushStreamId(streamId, params.sender, params.recipient);
     }
 
-    function createWithDurationsLPG(
+    function createWithTimestampsLPG(
         uint256 timeJumpSeed,
-        Lockup.CreateWithDurations memory params,
+        Lockup.CreateWithTimestamps memory params,
         uint40 duration,
```
Comment (smol-ninja) [inline lockup/tests/invariant/handlers/LockupCreateHandler.sol]: Why do we need `duration` since `params` has both start and end time?

**Review by smol-ninja — CHANGES_REQUESTED** [review]
Looks good to me. I have left my review below:

File: lockup/tests/integration/concrete/lockup-price-gated/create-with-timestamps-lpg/createWithTimestampsLPG.tree (hunk 3 above)
Comment (andreivladbrg) [reply lockup/tests/integration/concrete/lockup-price-gated/create-with-timestamps-lpg/createWithTimestampsLPG.tree:1]: > Since there is only one function for LPG, what if we create a full BTT tree for this similar to what we had in `createWithDUrationsLPG.tree`?

sounds good, we should add it

File: lockup/src/interfaces/ISablierBatchLockup.sol (hunk 1 above)
Comment (andreivladbrg) [reply lockup/src/interfaces/ISablierBatchLockup.sol:117]: is relevant tho for one who's the reading documentation of `createWithTimestampsLPG`?

File: lockup/tests/invariant/handlers/LockupCreateHandler.sol (hunk 5 above)
Comment (andreivladbrg) [reply lockup/tests/invariant/handlers/LockupCreateHandler.sol]: yeah, we don't need it

File: lockup/src/interfaces/ISablierBatchLockup.sol (hunk 1 above)
Comment (smol-ninja) [reply lockup/src/interfaces/ISablierBatchLockup.sol:117]: No, but to people who are looking for `createWithDurationsLPG` function.

File: lockup/tests/integration/concrete/lockup-price-gated/create-with-timestamps-lpg/createWithTimestampsLPG.tree (hunk 3 above)
Comment (andreivladbrg) [reply lockup/tests/integration/concrete/lockup-price-gated/create-with-timestamps-lpg/createWithTimestampsLPG.tree:1]: actually we inherit the `CreateWithTimestamps_Integration_Concrete_Test` so those tests are already covered

https://github.com/sablier-labs/lockup/blob/7033273efe4302b4b579d8f7a30b780bb030562c/lockup/tests/integration/concrete/lockup-price-gated/create-with-timestamps-lpg/createWithTimestampsLPG.t.sol#L17

Comment (smol-ninja) [reply lockup/tests/integration/concrete/lockup-price-gated/create-with-timestamps-lpg/createWithTimestampsLPG.tree:1]: Oh, you are right.

File: lockup/src/interfaces/ISablierBatchLockup.sol (hunk 1 above)
Comment (andreivladbrg) [reply lockup/src/interfaces/ISablierBatchLockup.sol:117]: alright

Comment (andreivladbrg) [reply lockup/src/interfaces/ISablierBatchLockup.sol:117]: db8752897e3c650babee13fa9c2654717b84d5db

File: lockup/src/interfaces/ISablierLockupPriceGated.sol (hunk 2 above)
Comment (andreivladbrg) [reply lockup/src/interfaces/ISablierLockupPriceGated.sol:63]: db8752897e3c650babee13fa9c2654717b84d5db

File: lockup/tests/integration/concrete/lockup-price-gated/streamed-amount-of/streamedAmountOf.t.sol (hunk 4 above)
Comment (andreivladbrg) [reply lockup/tests/integration/concrete/lockup-price-gated/streamed-amount-of/streamedAmountOf.t.sol]: db8752897e3c650babee13fa9c2654717b84d5db

File: lockup/tests/invariant/handlers/LockupCreateHandler.sol (hunk 5 above)
Comment (andreivladbrg) [reply lockup/tests/invariant/handlers/LockupCreateHandler.sol]: db8752897e3c650babee13fa9c2654717b84d5db

**Comment by andreivladbrg** [comment]
@smol-ninja pushed a new commit - if it looks good could mark the comments as resolved and approve the PR?

**Review by smol-ninja — APPROVED** [review]
